File: src/main/java/io/bindernews/thegrackle/patches/SingleCardViewPopupPatches.java
package io.bindernews.thegrackle.patches;

import basemod.ReflectionHacks;
import com.evacipated.cardcrawl.modthespire.lib.SpirePatch;
import com.megacrit.cardcrawl.cards.AbstractCard;
import com.megacrit.cardcrawl.cards.CardGroup;
import com.megacrit.cardcrawl.screens.SingleCardViewPopup;
import io.bindernews.thegrackle.Events;
import io.bindernews.thegrackle.interfaces.SvcChangeCardEvent;
import lombok.val;

@SuppressWarnings("unused")
public class SingleCardViewPopupPatches {

    @SpirePatch(clz = SingleCardViewPopup.class, method = "open", paramtypez = {AbstractCard.class, CardGroup.class})
    public static class CatchOpen {
        public static void Postfix(SingleCardViewPopup __instance, AbstractCard card, CardGroup group) {
            Events.svcCardChange().emit(new SvcChangeCardEvent(card, group, true));
        }
    }

    @SpirePatch(clz = SingleCardViewPopup.class, method = "close")
    public static class CatchClose {
        public static void Prefix(SingleCardViewPopup __instance) {
            //noinspection UnnecessaryLocalVariable
            val scv = __instance;
            val card = (AbstractCard) ReflectionHacks.getPrivate(scv, scv.getClass(), "card");
            val group = (CardGroup) ReflectionHacks.getPrivate(scv, scv.getClass(), "group");
            Events.svcCardChange().emit(new SvcChangeCardEvent(card, group, false));
        }
    }
}

File: src/main/kotlin/io/bindernews/thegrackle/patches/SingleCardViewPopupPatches.kt
@file:Suppress("UNUSED_PARAMETER", "FunctionName", "unused")

package io.bindernews.thegrackle.patches

import basemod.ReflectionHacks.getPrivate
import com.evacipated.cardcrawl.modthespire.lib.SpirePatch
import com.megacrit.cardcrawl.cards.AbstractCard
import com.megacrit.cardcrawl.cards.CardGroup
import com.megacrit.cardcrawl.screens.SingleCardViewPopup
import io.bindernews.thegrackle.Events.svcCardChange
import io.bindernews.thegrackle.api.SvcChangeCardEvent


class SingleCardViewPopupPatches {
    @SpirePatch(clz = SingleCardViewPopup::class, method = "open", paramtypez = [AbstractCard::class, CardGroup::class])
    object CatchOpen {
        @JvmStatic fun Postfix(__instance: SingleCardViewPopup, card: AbstractCard, group: CardGroup) {
            svcCardChange.emit(SvcChangeCardEvent(card, group, true))
        }
    }

    @SpirePatch(clz = SingleCardViewPopup::class, method = "close")
    object CatchClose {
        @JvmStatic fun Prefix(__instance: SingleCardViewPopup) {
            val card = getPrivate<AbstractCard>(__instance, __instance.javaClass, "card")
            val group = getPrivate<CardGroup>(__instance, __instance.javaClass, "group")
            svcCardChange.emit(SvcChangeCardEvent(card, group, false))
        }
    }
}